[Concurrency] Async Let (#1338)
apple/swift-evolution: proposals/mmmm-async-let.md

Similarly to the `group.async()` function, the closure is `@Sendable` and `nonisolated`, meaning that it cannot access non-sendable state of the enclosing context. For example, it will result in a compile-time error, preventing a potential race condition, for a `async let` initializer to attempt mutating a closed-over variable:

```swift
var localText: [String] = ...
async let w = localText.removeLast() // error: mutation of captured var 'localText' in concurrently-executing code
```

The async let initializer may refer to any sendable state, same as any non-isolated sendable closure.

The initializer of a `async let` permits the omission of the `await` keyword if it is directly calling an asynchronous function, like this:

```swift
func order() async -> Order { ... }

async let o1 = await order()
// should be written instead as
async let o2 = order()
```

This is because by looking at the async let declaration, it is obvious that the right-hand side function will be used to initialize the left-hand side, by waiting on it. This is similar to single-expression `return` keyword omission, and also applies only to single expression initializers.

> It is illegal to declare a `spawn var`. This is due to the complex initialization that a `async let` represents, it does not make sense to allow further external modification of them. Doing so would tremendously complicate the understandability of such asynchronous code, and undermine potential optimizations by making it harder to make assumptions about the data-flow of the values.

Reviewer: ```suggestion
It is illegal to declare an `async var`. This is due to the complex initialization that a `async let` represents, it does not make sense to allow further external modification of them. Doing so would tremendously complicate the understandability of such asynchronous code, and undermine potential optimizations by making it harder to make assumptions about the data-flow of the values.
```

```swift

> spawn var x = nope() // error: 'spawn' can only be used with 'let' declarations

Reviewer: ```suggestion
async var x = nope() // error: 'async' can only be used with 'let' declarations
```

If we instead awaited on one of the values, e.g. the fast one (`f`) the emitted code would not need to implicitly cancel or await it, as this was already taken care of explicitly:

```swift
func go2() async {
  async let f = fast()
  async let s = slow()
  _ = await f
  return "nevermind..."
  // implicitly: cancels s
  // implicitly: awaits s
}
```

The duration of the `go2()` call remains the same, it is always `time(go2) == max(time(f), time(s))`.

Special attention needs to be given to the `async let _ = ...` form of declarations. This form is interesting because it creates a child-task of the right-hand-side initializer, however it actively chooses to ignore the result. Such a declaration (and the associated child-task) will run and be awaited-on implicitly, as the scope it was declared in is about to exit — the same way as an unused `async let` declaration would be.

> > It may be interesting for a future proposal to explore the viability of sugar to `spawn voidReturningFunction()` directly, as a `Void` returning function may often not necessarily want to be awaited on for control-flow reasons, as variables of interesting types would be.

Reviewer: Does `spawn voidReturningFunction()` need to change?

Author: let me remove this actually, it's superseded by `async {}` which is an update to structured concurrency in https://github.com/apple/swift-evolution/pull/1345

    print("It's Friday!")
  } else {
    _ = try await registered
    // registration didn't throw, let's greet the guest!  
    print("Any other day of the week.")
  }
 
  return "Hello \(name)!"
}
```

By just looking at this code, it is not clear if the programmer _intentionally_ did not await on the registration on the `isFriday` branch, or if it is a real mistake, and the check must always throw. In other words, is this a place where everyone is let in on Fridays, but on other days only registered members are allowed on? :thinking: The code does not help us understand the real intent, and we would have to resort to code comments.

It might be better if it were _enforced_ by the compiler to _always_ (unless throwing or returning) to have to await on all `async let` declarations. E.g. in the example above, we could detect that there exist branches on which the registered was not awaited on, and signal this as an error to the programmer, who would have to:

- either fill in the appropriate `try await registered` inside the `isFriday` branch, or
- move the `async let registered` declaration into the `else` branch — we indeed only perform this check on non-Fridays.

> This rule might be too cumbersome for some code though, so perhaps this warrants a future extension where it is possible to require `@exhaustiveSpawnLetWaiting` on function level, to enforce that async lets are awaited on all code paths.

Reviewer: ```suggestion
This rule might be too cumbersome for some code though, so perhaps this warrants a future extension where it is possible to require `@exhaustiveAsyncLetWaiting` on function level, to enforce that async lets are awaited on all code paths.
```

Author: thanks; eh I'm terrible at those mass renames it seems

A `async let` declaration, in reality, is not just a plain sugar-syntax for task groups, and can make use of additional known-at-compile-time structure of the declared tasks. For example, it is possible to avoid heap allocations for small enough `async let` child tasks, avoid queues and other mechanisms which a task group must make use of to implement its "by completion order" yielding of values out of `next()`. 

This comes at a price though, async let declarations are less flexible than groups, and this is what we'll explore in this section.

Specifically, `async let` declarations are not able to express dynamic numbers of tasks executing in parallel, like this group showcases:

```swift
func toyParallelMap<A, B>(_ items: [A], f: (A) async -> B) async -> [B] { 
  return await withTaskGroup(of: (Int, B).self) { group in
    var bs = [B?](repeating: nil, count: items.count)
    
    // spawn off processing all `f` mapping functions in parallel
    // in reality, one might want to limit the "width" of these
    for i in items.indices { 

>       group.spawn { (i, await f(items[i])) }

Reviewer: ```suggestion
      group.async { (i, await f(items[i])) }
```
